[Clang] Implement the core language parts of P2786 - Trivial relocation

cor3ntin opened the pull request:
This adds

- The parsing of `trivially_relocatable_if_eligible`, `replaceable_if_eligible` keywords
- `__builtin_trivially_relocate`, implemented in terms of memmove. In the future this should
     - Add the appropriate start/end lifetime markers that llvm does not have (`start_lifetime_as`)
     - Add support for ptrauth when that's upstreamed

- the `__builtin_is_cpp_trivially_relocatable` and `__builtin_is_replaceable` traits


Fixes #127609

cor3ntin committed:
[Clang][WIP] Trivial relocation

cor3ntin committed:
add sema tests for __builtin_trivially_relocate

cor3ntin committed:
rename __is_cpp_trivially_relocatable

cor3ntin committed:
add docs and fix variable names in SemaChecking

cor3ntin committed:
restore assert, add codegen test

cor3ntin commented:
- [ ] extension warnings for the keywords
- [ ] changelog
- [ ] cxx_status page

zyn0217 commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12427,6 +12427,11 @@ def err_builtin_invalid_arg_type: Error <
   "a vector of floating points}1 (was %2)">;
 
+def err_builtin_trivially_relocate_invalid_arg_type: Error <
+  "first%select{||| and second}0 argument%select{|||s}0 to "
```
Can we use `%plural` or `%s` format specifier? Moreover, we might want to replace `%select` with `%enum_select`

cor3ntin replied:
Not really, we can't specify enum_select twice here, and plural does not apply

zyn0217 commented on clang/lib/AST/Type.cpp:
```diff
@@ -2862,6 +2862,30 @@ bool QualType::isTriviallyRelocatableType(const ASTContext &Context) const {
+  QualType BaseElementType = Context.getBaseElementType(*this);
+  if (BaseElementType->isIncompleteType())
+    return false;
+  else if (BaseElementType->isScalarType())
+    return true;
+  else if (const auto *RD = BaseElementType->getAsCXXRecordDecl())
+    return RD->isTriviallyRelocatable();
```
```suggestion
  if (BaseElementType->isScalarType())
    return true;
  if (const auto *RD = BaseElementType->getAsCXXRecordDecl())
    return RD->isTriviallyRelocatable();
```

zyn0217 commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -7258,6 +7261,217 @@ void Sema::CheckCompletedCXXClass(Scope *S, CXXRecordDecl *Record) {
+}
+
+void Sema::CheckCXX2CReplaceable(CXXRecordDecl *D) {
+  if (!D->hasDefinition() || D->isInvalidDecl())
+    return;
```
Is there any chance we end up here with D being incomplete, despite calling this function from `CheckCompletedCXXClass`?

zyn0217 commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -757,7 +757,7 @@ static void InitializeCPlusPlusFeatureTestMacros(const LangOptions &LangOpts,
   }
 
-  // We provide those C++23 features as extensions in earlier language modes, so
+  // We provide those C++2b features as extensions in earlier language modes, so
```
Why c++2b?

cor3ntin committed:
address comments, fix tests

cor3ntin committed:
address comments

cor3ntin committed:
format

cor3ntin committed:
add extension warning and tests

cor3ntin committed:
changelog, docs, C++03 tests

Sirraide commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -1875,6 +1875,54 @@ static ExprResult BuiltinIsWithinLifetime(Sema &S, CallExpr *TheCall) {
+  }
+
+  Expr *SizeExpr = TheCall->getArg(2);
+  ExprResult Size = S.DefaultLvalueConversion(SizeExpr);
```
On that note, don’t we also need to perform `DefaultLvalueConversion` on the two pointer arguments?

Sirraide commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -7258,6 +7261,221 @@ void Sema::CheckCompletedCXXClass(Scope *S, CXXRecordDecl *Record) {
+         (Implicit ? HasDefaultedCopyAssignment : true);
+}
+
+void Sema::CheckCXX2CTriviallyRelocatable(CXXRecordDecl *D) {
```
When it comes to this function, I’m having to think pretty hard to try and pair the code here w/ the wording on the standard; imo it would help if you could add some comments that briefly show which sections in this function correspond to what condition(s) in the standard. 

Also, I think factoring most of this out into either a lambda or a separate function so we can early-return whenever `IsTriviallyRelocatable` is set to `false` here would aid readability as well.

Sirraide replied:
Also, I wonder if having separate functions for e.g. `isDefaultMovable()` and `isEligibleForTrivialRelocation()` might help here.

Sirraide commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -7258,6 +7261,221 @@ void Sema::CheckCompletedCXXClass(Scope *S, CXXRecordDecl *Record) {
+  D->setIsTriviallyRelocatable(IsTriviallyRelocatable);
+}
+
+void Sema::CheckCXX2CReplaceable(CXXRecordDecl *D) {
```
This one is far less complicated but I’d also prefer if we could do `return false` instead of ` IsReplaceable = false` here (maybe just an immediately-invoked lambda in the initialiser of `IsReplaceable` is enough for this function?)

Sirraide commented on clang/test/SemaCXX/cxx2c-trivially-relocatable.cpp:
```diff
@@ -0,0 +1,272 @@
+    void* b;
+    int c[3];
+    Trivial d[3];
+    NonRelocatable& e = NonRelocatable_g;
```
Maybe add some `const`/`volatile` members too.

Sirraide commented on clang/test/SemaCXX/cxx2c-trivially-relocatable.cpp:
```diff
@@ -0,0 +1,272 @@
+static_assert(!__builtin_is_cpp_trivially_relocatable(int&));
+static_assert(!__builtin_is_cpp_trivially_relocatable(Trivial&));
+static_assert(__builtin_is_cpp_trivially_relocatable(const Trivial));
+static_assert(__builtin_is_cpp_trivially_relocatable(Trivial[1]));
```
Are VLA’s trivially relocatable? Either way, we probably want a test for that too

cor3ntin replied:
I don't think we can spell that

Sirraide replied:
Hmm, does this not do what I think it does?
```c++
int n = 4;
static_assert(__builtin_is_cpp_trivially_relocatable(int[n]));
```

Sirraide commented on clang/test/SemaCXX/cxx2c-trivially-relocatable.cpp:
```diff
@@ -0,0 +1,272 @@
+struct UserDtr {
+    ~UserDtr();
+};
+
```
A test w/ a deleted dtor would be nice

Sirraide commented on clang/test/SemaCXX/cxx2c-trivially-relocatable.cpp:
```diff
@@ -0,0 +1,272 @@
+
+template <typename T>
+struct WithVBaseExplicit replaceable_if_eligible : virtual T{};
+static_assert(__builtin_is_replaceable(WithVBaseExplicit<S<int>>)); // expected-error {{failed}}
```
```suggestion
static_assert(!__builtin_is_replaceable(WithVBaseExplicit<S<int>>));
```
Any reason we’re not just doing this here?

Sirraide commented on clang/test/SemaCXX/cxx2c-trivially-relocatable.cpp:
```diff
@@ -0,0 +1,272 @@
+
+void test__builtin_trivially_relocate() {
+    struct S{ ~S();};
+    struct R {};
```
Maybe add a test w/ `__builtin_trivially_relocate()` in a template, particularly w/ a dependent src/dest argument.

cor3ntin commented on clang/lib/AST/Type.cpp:
```diff
@@ -2862,6 +2862,30 @@ bool QualType::isTriviallyRelocatableType(const ASTContext &Context) const {
+}
+
+bool QualType::isReplaceableType(const ASTContext &Context) const {
+  if (isConstQualified())
```
Nope, volatile types are replaceable

cor3ntin committed:
rewrite everything

frederick-vs-ja commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -1826,6 +1827,12 @@ The following type trait primitives are supported by Clang. Those traits marked
   source object on the floor. This is true of trivial types and types which
   were made trivially relocatable via the ``clang::trivial_abi`` attribute.
+* ``__builtin_is_cpp_trivially_relocatable`` (C++): Returns true if and object
+  is trivially relocatable, as defined by the C++26 standard.
```
[meta.unary.prop] and/or [tab:meta.unary.prop] will be suitable.

cor3ntin replied:
Sure, done. note that we don;t do that for anything else though

erichkeane replied:
TBH we probably should.  I asked for it because 'as defined by the C++26 standard' is pretty vague.

cor3ntin committed:
address more feedback

cor3ntin commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -127,6 +127,33 @@ class AccessSpecDecl : public Decl {
+  SourceLocation getLocation() const { return Loc; }
+
+private:
+  SourceLocation Loc;
```
It used to have more. But I kept it because I felt a string type was cleaner than a bunch of lone SourceLocations. I'd rather keep it, unless you feel strongly about it

erichkeane replied:
It seems like a lot of 'overhead' mentally to have a type here that all it does is contain a single thing?  But I guess I don't feel super strongly.

philnik777 commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -1826,6 +1827,12 @@ The following type trait primitives are supported by Clang. Those traits marked
   functionally equivalent to copying the underlying bytes and then dropping the
   source object on the floor. This is true of trivial types and types which
   were made trivially relocatable via the ``clang::trivial_abi`` attribute.
+* ``__builtin_is_cpp_trivially_relocatable`` (C++): Returns true if an object
+  is trivially relocatable, as defined by the C++26 standard [meta.unary.prop].
+  Note that the caller code should ensure that if the object is polymorphic,
+  the dynamic type is of the most derived type.
```
Why do we introduce another builtin for this instead of fixing the existing one?

cor3ntin replied:
I'd like a built in that has the standard behavior.
I think there is an interesting question as to whether the existing builtin should be modified to return true
if either a type is trivial_abi or relocatable by the standard. I think this would make sense

philnik777 replied:
Wouldn't it make a lot more sense to deprecate marking types as `trivial_abi` that aren't trivially relocatable? AFIAICT that's a requirement anyways, which is why it has been added in the first place. Then we can just say "__is_trivially_relocatable returns only true for types that are according to the standard" instead of introducing a new builtin just for the old builtin to turn out to be more useful in the end anyways.

cor3ntin replied:
I would be happy doing that but I'm afraid it would break existing code.
Right now 

__is_cpp_trivially_relocatable => standard behavior
__is_trivially_relocatable => standard behavior + trivial_abi

philnik777 replied:
I'm not sure I understand what breakage you're afraid of. Sure, a deprecation warning would break stuff with `-Werror`, but that's nothing new. I'm not saying we should make `__is_trivially_relocatable` return false if only `[[clang::trivial_abi]]` is added. I'm saying we should fix the users code and warn (and eventually error) if they're not giving us the guarantee. I just don't see when we'd ever want to use `__is_cpp_trivially_relocatable` over `__is_trivially_relocatable`.

cor3ntin replied:
`std::is_trivially_relocatable` should *not* return true for a type that is not relocatable per the standard.

This gives us a few options, none of them great
- Introduce a new trait so that we can model the standard behavior
- Modify the existing trait to ignore `[[clang::trivial_abi]]`  which _will_ break code
- Be not conforming

Having a warning does not solve the conformance issue.
I agree with you that we probably want
 - Restrict the set of types that can be  `[[clang::trivial_abi]]`  (ie warn if  `[[clang::trivial_abi]]`  is added to a non-relocatable type)
 - Converge to a single trait by deprecating the old trait.

erichkeane replied:
Corentin is right.  Any attempt we make to modify `__is_trivially_relocatable` is going to either be non-standard or break `[[clang::trivial_abi]]`.

SO, we'd want to choose which we'd want to do.  I very strongly don't believe we can be non-standard here, this is an important standardized feature, and `clang::trivial_abi` isn't any better.

I am ok with us modifying `clang::trivial_abi` over a number of releases(like 2-3 releases!) to make it be the same as the standards feature, but only if we can come to an agreement with the folks who are already using it.

SO in short, I think combining these builtins right now is untenable.  Long-term, it is  possible but only if we are willing/able to significantly break `clang::trivial_abi`.

cor3ntin replied:
I made an issue for that https://github.com/llvm/llvm-project/issues/128725

philnik777 replied:
> `std::is_trivially_relocatable` should _not_ return true for a type that is not relocatable per the standard.

Why? It's our extension to define. Am I missing something that makes annotating a type `[[clang::trivial_abi]]` instead of `trivially_relocatable_if_eligible [[clang::trivial_abi]]` significantly different? AFAIK having a type marked `[[clang::trivial_abi]]` requires it to be trivially relocatable.

cor3ntin replied:
> AFAIK having a type marked [[clang::trivial_abi]] requires it to be trivially relocatable.

It's not exactly the same semantics.

In particular, the standard traits allow for polymorphic types, any code that would do a memcpy without checking for that would lead to nasty UB.

How special members contribute to relocatability is also different, the standard requires non-deleted assignment.
That would also break code or be forever non-conforming and portable.

cor3ntin committed:
Add comments

cor3ntin committed:
Fix volatile test

Fznamznon commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -7258,6 +7261,228 @@ void Sema::CheckCompletedCXXClass(Scope *S, CXXRecordDecl *Record) {
+  return !Dtr->isDeleted();
+}
+
+static bool hasDeletedDestructor(CXXRecordDecl *D) {
```
Does it make sense to make `hasDeletedDestructor` a method of `CXXRecordDecl`? Maybe it will end up bein useful around codebase.

cor3ntin replied:
Done

Fznamznon commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -3808,13 +3868,39 @@ void Parser::ParseCXXMemberSpecification(SourceLocation RecordLoc,
-      if (Specifier == VirtSpecifiers::VS_None)
-        break;
+      if (Specifier == VirtSpecifiers::VS_None) {
+        if (isCXX2CTriviallyRelocatableKeyword(Tok)) {
+          if (TriviallyRelocatable.isSet()) {
+            auto Skipped = Tok;
+            ConsumeToken();
+            Diag(Skipped, diag::err_duplicate_class_relocation_specifier)
+                << /*trivial_relocatable*/ 0
+                << TriviallyRelocatable.getLocation();
+          } else {
+            ParseCXX2CTriviallyRelocatableSpecifier(TriviallyRelocatable);
+          }
+          continue;
+        } else if (isCXX2CReplaceableKeyword(Tok)) {
+          if (Replacable.isSet()) {
+            auto Skipped = Tok;
+            ConsumeToken();
+            Diag(Skipped, diag::err_duplicate_class_relocation_specifier)
+                << /*replaceable*/ 1 << Replacable.getLocation();
+          } else {
+            ParseCXX2CReplaceableSpecifier(Replacable);
+          }
+          continue;
```
I have a feeling that if TriviallyRelocatableSpecifier/ReplaceableSpecifier were not templated and instead had some enum "Kind" field, the following two almost identical branches could be rewritten into:
```
BasicRelocatableOrReplaceableClassSpecifier ToComplain;


if (isCXX2CTriviallyRelocatableKeyword(Tok))
  ToComplain = TriviallyRelocatable;
else if (isCXX2CReplaceableKeyword(Tok))
  ToComplain = Replacable;

if (ToComplain.isSet()) {
    auto Skipped = Tok;
    ConsumeToken();
    Diag(Skipped, diag::err_duplicate_class_relocation_specifier)
    << /*trivial_relocatable*/ ToComplain.getKind()
    << TriviallyRelocatable.getLocation();
} else if (ToComplain.isNotEmpty()) {
  (ToComplain.getKind)? ParseCXX2CTriviallyRelocatableSpecifier(ToComplain) : ParseCXX2CReplaceableSpecifier(ToComplain);
}

```
that is an entirely optional suggestion but that amount of else/if/{} for a technically same code just makes it look sad.

Fznamznon commented on clang/test/Parser/cxx2c-trivially-relocatable.cpp:
```diff
@@ -0,0 +1,42 @@
+
+
+class A trivially_relocatable_if_eligible {};
+// cxx11-warning@-1 {{'trivially_relocatable_if_eligible' keyword is a C++2c extension}}
```
I'm just curious, why don't we say and test C++26 instead of C++2c?

cor3ntin committed:
* Use overload on special members
* fix tests (class with a const member are not implicitly relocatable)
* Factorize some code to only compute default-movable once

cor3ntin committed:
Make hasDeletedDestructor a member

cor3ntin committed:
__is_trivially_relocatable is true for c++26 relocatable types

cor3ntin committed:
Remove the specifier wrapper, stash the location in attributes

zyn0217 commented on clang/include/clang/AST/Type.h:
```diff
@@ -1129,6 +1129,10 @@ class QualType {
   /// Return true if this is a trivially relocatable type.
   bool isTriviallyRelocatableType(const ASTContext &Context) const;
 
+  bool isCppTriviallyRelocatableType(const ASTContext &Context) const;
```
FYI there are `QualType::isCXX{98,11}PODType` around, so probably `isCXXTriviallyRelocatableType` for consistency?

zyn0217 commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -3808,13 +3867,38 @@ void Parser::ParseCXXMemberSpecification(SourceLocation RecordLoc,
+            ParseCXX2CTriviallyRelocatableSpecifier(TriviallyRelocatable);
+          }
+          continue;
+        } else if (isCXX2CReplaceableKeyword(Tok)) {
```
there should be some clang-tidy warnings suggesting dropping these `else`s after `continue`...?

cor3ntin committed:
Defer the computation of relocatability/replaceability to
when the trait is queried, in order to avoid a performance
regression.

Unfortunately, this move the queries from ASTContext to Sema.

cor3ntin committed:
reduce destructor lookups

cor3ntin committed:
cleanup

cor3ntin committed:
remove dead code

erichkeane commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -624,6 +624,20 @@ class ASTContext : public RefCountedBase<ASTContext> {
+                                     CXXRecordDeclRelocationInfo);
+
+private:
+  llvm::DenseMap<const CXXRecordDecl *, CXXRecordDeclRelocationInfo>
```
With the 'value' here only being 2 bits of information, I find myself wondering if we could switch `CXXRecordDeclRelocationInfo` to an enum-as-flags, and do something with `PointerIntPair` here.  Then store these either as a sorted vector, or a 'set' in some way?  Perhaps have a look around? 

Also, do we typically do caching like this in ASTContext? I thought we did this in Sema more often.  

Also-also: does this play well with deserialization?

cor3ntin replied:
> With the 'value' here only being 2 bits of information, I find myself wondering if we could switch CXXRecordDeclRelocationInfo to an enum-as-flags, and do something with PointerIntPair here. Then store these either as a sorted vector, or a 'set' in some way? Perhaps have a look around?

I did think about that but I don't see how it would not massively penalize lookup.
(Note that there is an existing defect that we don't cache `__bitwise_copyable`, and maybe we'd want to use the same map for that)

> Also, do we typically do caching like this in ASTContext? I thought we did this in Sema more often.

I mean, the fact we have a type property depending on Sema is pretty novel (and I hate it!) - but yeah, all of the other cached properties are stored in ASTContext and that seem like the most logical to me


> Also-also: does this play well with deserialization?
 
Given the property is always on the definition, yes?

cor3ntin committed:
Add VLA tests

cor3ntin committed:
Format

cor3ntin committed:
Make sure __is_trivially_relocatable is false for polymorphic types

cor3ntin committed:
Do not yet set the featurte test macro

cor3ntin committed:
format

ldionne reviewed (commented):
I would like this patch to also deprecate the old builtin. I think the existence of the old builtin is going to cause significant confusion and risk introducing bugs. I always imagined that the old builtin would in fact be "fixed in place" once the Standard introduced the notion of TR, but I understand that introducing the Standard thing as a new name is probably better for backwards compatibility reasons. That being said, I would definitely want the old one to be deprecated with an explanation the very moment we introduce the new, Standards-conforming thing.

erichkeane commented:
> I would like this patch to also deprecate the old builtin. I think the existence of the old builtin is going to cause significant confusion and risk introducing bugs. I always imagined that the old builtin would in fact be "fixed in place" once the Standard introduced the notion of TR, but I understand that introducing the Standard thing as a new name is probably better for backwards compatibility reasons. That being said, I would definitely want the old one to be deprecated with an explanation the very moment we introduce the new, Standards-conforming thing.

I think this is a very reasonable request.

cor3ntin commented:
@ldionne @erichkeane are you happy with a follow-up pr (within a few days of landing this) ?

erichkeane commented:
I think the promise to do the deprecation 'soon' (and by 'soon' it needs to be before the next release IMO, else we need to revert this, as releasing this without the diagnostic is, IMO, not tennable) is acceptable.

cor3ntin committed:
types with ptrauth are not relocatable for now

cor3ntin committed:
Fix ptrauth tests

cor3ntin committed:
format
